<a href="https://colab.research.google.com/github/Tessellate-Imaging/monk_v1/blob/master/study_roadmaps/2_transfer_learning_roadmap/2_acceptable_data_types/3)%20Multi%20label%20-%20Data%20with%20labels%20in%20CSV%20file.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Goals

### 1. Understand what multi label image classifications is
        - One or more than one label per image

### 2. Understand multi label - csv type data structure

## Table of Contents


## [0. Install](#0)


## [1. Load train data - csv type - Multi Label](#1)


## [2. Understand experiment data-summary](#2)


## [3. Load val data along with train data](#3)


## [4. Understand experiment data-summary](#4)

# Dataset Details
- Credits: https://www.kaggle.com/c/planet-understanding-the-amazon-from-space

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=12ia20erQXTmJTk1zj1pZ_VeCPViyAQvm' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=12ia20erQXTmJTk1zj1pZ_VeCPViyAQvm" -O satellite_cls.zip && rm -rf /tmp/cookies.txt

In [ ]:
! unzip -qq satellite_cls.zip

<a id='0'></a>
# Install Monk

## Using pip (Recommended)

  - colab (gpu) 
      - All bakcends: `pip install -U monk-colab`
      

  - kaggle (gpu) 
      - All backends: `pip install -U monk-kaggle`
      

  - cuda 10.2	
      - All backends: `pip install -U monk-cuda102`
      - Gluon bakcned: `pip install -U monk-gluon-cuda102`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda102`
      - Keras backend: `pip install -U monk-keras-cuda102`
      

  - cuda 10.1	
      - All backend: `pip install -U monk-cuda101`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda101`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda101`
	  - Keras backend: `pip install -U monk-keras-cuda101`
      

  - cuda 10.0	
      - All backend: `pip install -U monk-cuda100`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda100`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda100`
	  - Keras backend: `pip install -U monk-keras-cuda100`
      

  - cuda 9.2	
      - All backend: `pip install -U monk-cuda92`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda92`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda92`
	  - Keras backend: `pip install -U monk-keras-cuda92`
      

  - cuda 9.0	
      - All backend: `pip install -U monk-cuda90`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda90`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda90`
	  - Keras backend: `pip install -U monk-keras-cuda90`
      

  - cpu 		
      - All backend: `pip install -U monk-cpu`
	  - Gluon bakcned: `pip install -U monk-gluon-cpu`
	  - Pytorch backend: `pip install -U monk-pytorch-cpu`
	  - Keras backend: `pip install -U monk-keras-cpu`

## Install Monk Manually (Not recommended)
 
### Step 1: Clone the library
 - git clone https://github.com/Tessellate-Imaging/monk_v1.git
 
 
 
 
### Step 2: Install requirements 
 - Linux
     - Cuda 9.0
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu90.txt`
     - Cuda 9.2
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu92.txt`
     - Cuda 10.0
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu100.txt`
     - Cuda 10.1
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu101.txt`
     - Cuda 10.2
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu102.txt`
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Linux && pip install -r requirements_cpu.txt`
 
 
  - Windows
     - Cuda 9.0 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu90.txt`
     - Cuda 9.2 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu92.txt`
     - Cuda 10.0 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu100.txt`
     - Cuda 10.1 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu101.txt`
     - Cuda 10.2 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu102.txt`
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cpu.txt`
         
         
  - Mac
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Mac && pip install -r requirements_cpu.txt`
         
         
  - Misc
      - Colab (GPU)
          - `cd monk_v1/installation/Misc && pip install -r requirements_colab.txt`
      - Kaggle (GPU)
          - `cd monk_v1/installation/Misc && pip install -r requirements_kaggle.txt`
 
 
 
### Step 3: Add to system path (Required for every terminal or kernel run)
 - `import sys`
 - `sys.path.append("monk_v1/");`

# Imports

In [1]:
#Using gluon backend 

# When installed using pip
from monk.gluon_prototype import prototype


# When installed manually (Uncomment the following)
#import os
#import sys
#sys.path.append("monk_v1/");
#sys.path.append("monk_v1/monk/");
#from monk.gluon_prototype import prototype

## Creating and managing experiments
    - Provide project name
    - Provide experiment name

In [7]:
gtf = prototype(verbose=1);
gtf.Prototype("Dataset-Type", "exp-csv-multi-label");

Mxnet Version: 1.5.0

Experiment Details
    Project: Dataset-Type
    Experiment: exp-csv-multi-label
    Dir: /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.4_multilabel/workspace/Dataset-Type/exp-csv-multi-label/



### This creates files and directories as per the following structure
    
    
    workspace
        |
        |--------Dataset-Type 
                        |
                        |
                        |-----exp-foldered 
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)

<a id='1'></a>
# Load train data - foldered type

## Quick mode training

    - Using Default Function
        - dataset_path
        - model_name
        - num_epochs
        
        
## Dataset folder structure

    parent_directory
        |
        |
        |------train (folder can be named anything)
                |------img1.jpg
                |------img2.jpg
                |------.... (and so on)
        |
        |------train_labels.csv (file can be named anything)
        
        
    train_labels.csv has 2 columns (column names could be anything)
        - ID                 | Labels
          img1.jpg           |  agriculture cloudy
          img2.jpg           |  forest hazy
          img3.jpg           |  road forest rains
          img4.jpg           |  city
                ..... (and so on)
                
### In similar fashion you can structure your own dataset

### Along with providing dataset_path you now provide path_to_csv

In [8]:
gtf.Default(dataset_path="satellite_cls/train-jpg",
            path_to_csv="satellite_cls/train.csv",
            model_name="resnet18_v1", 
            
            delimiter = " ",    # very important to mention how the labels are separated, via spaces, or commas, etc
            
            
            num_epochs=5);

Dataset Details
    Train path:     satellite_cls/train-jpg
    Val path:       None
    CSV train path: satellite_cls/train.csv
    CSV val path:   None
    Label Type:     multiple

Dataset Params
    Input Size:   224
    Batch Size:   4
    Data Shuffle: True
    Processors:   4
    Train-val split:   0.7
    Delimiter:    

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 28335
    Num val images:   12144
    Num classes:      17

Model Params
    Model name:           resnet18_v1
    Use Gpu:              True
    Use pretrained:       True
    Freeze base network:  True

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
    

<a id='2'></a>
# Understand experiment data-summary

    Dataset Details
        Train path:     monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train
        Val path:       None
        CSV train path: monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train.csv
        CSV val path:   None
        
        
### Train path is mentioned and Val path is None

### CSV Train path is mentioned unlike in foldered type dataset


    Dataset Parameters:
        Input Size:   224
        Batch Size:   4
        Shuffle:      True
        Processors:   4
        Num Classes:  2

###  Data input size is 224 (default setting)
###  Batch size is 4 (default setting)

   Dataset Numbers
    Num train images: 28335
    Num val images:   12144
    Num classes:      17
        
### Since val data is missing, train data is split into train and val with a ratio 70:30 rule
    - 70% of data being used for training
    - 30% of data being used for validation
    
### Auto detect number of classes

# Goals Completed

### 1. Understand what multi label image classifications is
        - One or more than one label per image

### 2. Understand multi label - csv type data structure